File: src/test/java/org/exampilistic/engines/junit/ExampilisticEngineTest.java
package org.exampilistic.engines.junit;

import org.exampilistic.engines.junit.javatests.JavaTestLoader;
import org.exampilistic.engines.junit.jsontests.JsonTestLoader;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.platform.engine.EngineDiscoveryRequest;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.UniqueId;

import java.util.ArrayList;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ExampilisticEngineTest {

	private JsonTestLoader jsonTestLoader;
	private JavaTestLoader javaTestLoader;
	private ExampilisticEngine engine;

	@BeforeEach
	public void setup() {
		this.jsonTestLoader = mock(JsonTestLoader.class);
		this.javaTestLoader = mock(JavaTestLoader.class);
		this.engine = new ExampilisticEngine(jsonTestLoader, javaTestLoader);
	}

	@Test
	public void providesEngineId() {
		assertThat(engine.getId()).isEqualTo("exampilistic");
	}

	@Test
	public void discoverAlwaysReturnsTestDescriptor() {
		TestDescriptor testDescriptor = discoverAllTests();

		assertThat(testDescriptor).isNotNull();
	}

	@Test
	public void discoverReturnsTestDescriptorWithUniqueIdAndCorrectDisplayName() {
		EngineDiscoveryRequest discoveryRequest = mock(EngineDiscoveryRequest.class);
		UniqueId uniqueId = mock(UniqueId.class);
		TestDescriptor testDescriptor = engine.discover(discoveryRequest, uniqueId);

		//TODO add non-null assumption once we have assertj-3.9.0
		assertThat(testDescriptor.getUniqueId()).isSameAs(uniqueId);
		assertThat(testDescriptor.getDisplayName()).isEqualTo("Exampilistic");
	}

	@Test
	public void discoversAllTestsFromJsonLoader() {
		TestDescriptor expectedTestDescriptor = mock(TestDescriptor.class);
		installExpectedTestDescriptorTo(expectedTestDescriptor, jsonTestLoader);

		TestDescriptor testDescriptor = discoverAllTests();

		//TODO add non-null assumption once we have assertj-3.9.0
		//Here be dragons: Cannot remove the extracting call, even though it does nothing, because otherwise the
		//method contains will not be found (some generics issue).
		assertThat(testDescriptor.getChildren()).extracting(x -> (TestDescriptor) x).contains(expectedTestDescriptor);
	}

	private TestDescriptor discoverAllTests() {
		EngineDiscoveryRequest discoveryRequest = mock(EngineDiscoveryRequest.class);
		UniqueId uniqueId = mock(UniqueId.class);
		return engine.discover(discoveryRequest, uniqueId);
	}

	private void installExpectedTestDescriptorTo(TestDescriptor expectedTestDescriptor, TestLoader jsonTestLoader) {
		ArrayList<TestDescriptor> testDescriptors = new ArrayList<>();
		testDescriptors.add(expectedTestDescriptor);
		when(this.jsonTestLoader.loadAllTests()).thenReturn(testDescriptors);
	}


	@Test
	public void discoversAllTestsFromJavaLoader() {
		TestDescriptor expectedTestDescriptor = mock(TestDescriptor.class);
		installExpectedTestDescriptorTo(expectedTestDescriptor, javaTestLoader);

		TestDescriptor testDescriptor = discoverAllTests();

		//TODO add non-null assumption once we have assertj-3.9.0
		//Here be dragons: Cannot remove the extracting call, even though it does nothing, because otherwise the
		//method contains will not be found (some generics issue).
		assertThat(testDescriptor.getChildren()).extracting(x -> (TestDescriptor) x).contains(expectedTestDescriptor);
	}
}

File: src/test/kotlin/org/exampilistic/engines/junit/ExampilisticEngineTest.kt
package org.exampilistic.engines.junit

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.platform.engine.UniqueId
import org.mockito.Mockito.mock
import java.lang.IllegalArgumentException

class ExampilisticEngineTest {

	lateinit var exampilisticEngine : ExampilisticEngine
	lateinit var uniqueId : UniqueId

	@BeforeEach
	fun setUp() {
		uniqueId = mock(UniqueId::class.java)
		exampilisticEngine = ExampilisticEngine()
	}


	@Test
	fun discover_returns_engine_descriptor_of_proper_type() {
		val descriptor = exampilisticEngine.discover(null, uniqueId)
		assertThat(descriptor).isInstanceOf(ExampilisticEngineDescriptor::class.java)
	}

	@Test
	fun discover_returns_engine_with_proper_unique_id() {
		val descriptor = exampilisticEngine.discover(null, uniqueId)
		assertThat(descriptor.uniqueId).isEqualTo(uniqueId)
	}

	@Test
	fun discover_throws_exception_when_uniqueId_is_missing() {
		val exception = assertThrows(IllegalArgumentException::class.java, { -> exampilisticEngine.discover(null, null) })
		assertThat(exception.message).isEqualTo("uniqueId is missing")
	}
}